README.rst
===========
notifierlib
===========

A library that implements a kind of fan-out pattern, sending messages to very different endpoints.
Extendable through the implementation of custom Channels.


* Documentation: http://notifierlib.readthedocs.io/en/latest/

Features
--------

Introduced concepts are:

* Channels
    A channel is a communication endpoint capable of sending some type of message exposing a "notify" method.

* Groups
    A group is a construct bringing channels together under a common entrypoint showing up as a method call of the main Notifier object.

* Notifier
    The main object bringing together channels as broadcast by default and exposing methods to register and unregister channels and add and remove groups.

The payload of the notification methods is by convention a dictionary with at least a "subject" and a "message" key with the appropriate values.
It is designed like this so it can be very easy to implement much more complex structures like templates without needed any domain knowledge on all the channels.

Each channel can implement their own template as in instantiation argument in their class and handle the interpolation of variables in the notify method.

An example could be :

.. code-block:: python

    from notifierlib.channels import Email
    from notifierlib import Notifier, Group

    import logging
    logging.basicConfig(level=logging.DEBUG)
    template='<b>{{subject}}</b> <i>{{message}}</i> <a href="http://google.com">link</a>.'

    email=Email('email',
                sender='sender@gmail.com',
                recipient='recipient@gmail.com',
                smtp_address='smtp.domain.com',
                username='smtp_username',
                password='smtp_password',
                tls=True,
                ssl=False,
                port=587,
                template=template,
                content='html')

    notifier=Notifier()
    notifier.register(email)
    notifier.broadcast(subject='this is a test of a template', message="""this is a nice and long message""")


The above would render the template with the provided values on the notify method of the email channel before the mail gets sent.




For a more detailed explanation please see the USAGE.rst file.


HISTORY.rst
.. :changelog:

History
-------

0.0.1 (18-09-2017)
---------------------

* First code creation


tests/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: __init__.py
#
# Copyright 2017 Costas Tyfoxylos
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#

"""
.. _Google Python Style Guide:
   http://google.github.io/styleguide/pyguide.html
"""

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = '''google'''
__date__ = '''18-09-2017'''
__copyright__ = '''Copyright 2017, Costas Tyfoxylos'''
__license__ = '''MIT'''
__maintainer__ = '''Costas Tyfoxylos'''
__email__ = '''<costas.tyf@gmail.com>'''
__status__ = '''Development'''  # "Prototype", "Development", "Production".


tests/test_notifierlib.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: test_notifierlib.py
#
# Copyright 2017 Costas Tyfoxylos
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#

"""
test_notifierlib
----------------------------------
Tests for `notifierlib` module.

.. _Google Python Style Guide:
   http://google.github.io/styleguide/pyguide.html

"""

from betamax.fixtures import unittest

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = '''google'''
__date__ = '''18-09-2017'''
__copyright__ = '''Copyright 2017, Costas Tyfoxylos'''
__credits__ = ["Costas Tyfoxylos"]
__license__ = '''MIT'''
__maintainer__ = '''Costas Tyfoxylos'''
__email__ = '''<costas.tyf@gmail.com>'''
__status__ = '''Development'''  # "Prototype", "Development", "Production".


class TestNotifierlib(unittest.BetamaxTestCase):

    def setUp(self):
        """
        Test set up

        This is where you can setup things that you use throughout the tests. This method is called before every test.
        """
        pass

    def tearDown(self):
        """
        Test tear down

        This is where you should tear down what you've setup in setUp before. This method is called after every test.
        """
        pass


notifierlib/notifierlib.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
# File: notifierlib.py
"""
Main module file

Put your classes here
"""

import logging
import abc

from stopit import ThreadingTimeout, TimeoutException
from Queue import Queue
from threading import Thread

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = 'plaintext'
__date__ = '''18-09-2017'''

# This is the main prefix used for logging
LOGGER_BASENAME = '''notifierlib'''
LOGGER = logging.getLogger(LOGGER_BASENAME)
LOGGER.addHandler(logging.NullHandler())

WORKERS = 3
TIMEOUT = 30


class Channel(object):
    __metaclass__ = abc.ABCMeta

    def __init__(self, name):
        self._logger = logging.getLogger('{base}.{suffix}'
                                         .format(base=LOGGER_BASENAME,
                                                 suffix=self.__class__.__name__)
                                         )
        self.name = name

    @abc.abstractmethod
    def notify(self):
        pass


class Group(object):
    def __init__(self, group_name='', *channels):
        self._logger = logging.getLogger('{base}.{suffix}'
                                         .format(base=LOGGER_BASENAME,
                                                 suffix=self.__class__.__name__)
                                         )
        self.name = group_name
        self._channels = [self._validate_channel(channel)
                          for channel in channels]
        self._queue = Queue()
        self._results = None

    @staticmethod
    def _validate_channel(channel):
        if not isinstance(channel, Channel):
            raise ValueError('The object is not a Channel')
        return channel

    def _start_workers(self):
        for _ in range(WORKERS):
            worker = Thread(target=self._worker, args=(self._queue,))
            worker.setDaemon(False)
            worker.start()

    def __call__(self, **kwargs):
        self.send(**kwargs)

    def send(self, **kwargs):
        self._results = []
        for channel in self._channels:
            self._queue.put((channel, kwargs))
        self._start_workers()
        self._logger.debug('Waiting for results')
        self._queue.join()
        self._logger.debug(('Result of notification: '
                            '{result}').format(result=self._results))
        return self._results

    def _worker(self, queue):
        while not queue.empty():
            channel, kwargs = queue.get()
            self._logger.debug(('Sending notification using channel: {channel} '
                                'with args:{args}').format(channel=channel.name,
                                                           args=kwargs))
            try:
                with ThreadingTimeout(TIMEOUT):
                    result = channel.notify(**kwargs)
                    self._results.append({channel.name: result})
            except TimeoutException:
                self._logger.error(('The worker reached the time limit '
                                    '({} secs)').format(TIMEOUT))
            except Exception:
                self._logger.exception(('Exception caught on sending on '
                                        'channel:{}').format(channel.name))
            queue.task_done()


class Notifier(object):
    def __init__(self):
        self._logger = logging.getLogger('{base}.{suffix}'
                                         .format(base=LOGGER_BASENAME,
                                                 suffix=self.__class__.__name__)
                                         )
        self.broadcast = Group('broadcast')

    @property
    def channels(self):
        return [channel.name for channel in self.broadcast._channels]  # noqa

    def register(self, *args):
        for channel in args:
            if not isinstance(channel, Channel):
                raise ValueError(('The object is not a Channel :'
                                  '[{}]').format(channel))
            if channel.name in self.channels:
                raise ValueError('Channel already registered')
            self.broadcast._channels.append(channel)  # noqa

    def unregister(self, *args):
        for channel in args:
            if channel.name not in self.channels:
                raise ValueError('Channel not registered')
            self.broadcast._channels = [ch for ch in self.broadcast._channels  # noqa
                                        if not ch.name == channel.name]

    def add_group(self, group):
        if not isinstance(group, Group):
            raise ValueError(('The object is not a Group :'
                              '[{}]').format(group))
        setattr(self, group.name, group)

    def remove_group(self, group):
        if not isinstance(group, Group):
            raise ValueError(('The object is not a Group :'
                              '[{}]').format(group))
        try:
            delattr(self, group.name)
            return True
        except AttributeError:
            self._logger.error('No such group :{}'.format(group.name))
            return False


notifierlib/channels/slack.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
# File: slack.py
"""Slack module file"""

import requests
from notifierlib.notifierlib import Channel
from jinja2 import Environment

__author__ = '''Oriol Fabregas <fabregas.oriol@gmail.com>'''
__docformat__ = 'plaintext'
__date__ = '''20-09-2017'''


class AuthenticationError(Exception):
    """If the response from the server is not successful in getting user id"""


class Slack(Channel):
    """
    Integration with Slack

    In a team, a channel is public by design, therefore the attribute 'private'
    is False by default. A private channel is just a group.

    Slack API understands a channel name as "#channel" format and a group as
    "group". Also, by default will reply to the whole channel.

    "as_user" is also implemented by default in this notification.

    +info https://api.slack.com/methods/chat.postMessage#channels
    """
    def __init__(self,
                 name,
                 token,
                 channel,
                 template=None,
                 private=False,
                 reply_broadcast=False):
        super(Slack, self).__init__(name)
        self.private = private
        self.channel = channel if self.private else '#{}'.format(channel)
        self.template = template
        self.site = 'https://slack.com/api'
        self.reply_broadcast = reply_broadcast
        self.__token = token

    def _get_user_id(self):
        url = '{site}/auth.test'.format(site=self.site)
        response = requests.post(url, data={'token': self.__token})
        if not response.json().get('ok'):
            message = ('Error getting user details from {url}.\n'
                       'Response: {res}\n').format(url=response.url,
                                                   res=response.content)
            self._logger.error(message)
            raise AuthenticationError
        return response.json().get('user_id')

    def notify(self, **kwargs):
        try:
            if self.template:
                body = Environment().from_string(self.template).render(**kwargs)
            else:
                body = kwargs.get('message')
            arguments = {'channel': self.channel,
                         'token': self.__token,
                         'text': body,
                         'reply_broadcast': self.reply_broadcast,
                         'as_user': self._get_user_id()}
            url = '{site}/chat.postMessage'.format(site=self.site)
            response = requests.post(url, data=arguments)
            if not response.json().get('ok'):
                message = ('Error while sending message to channel {} \n'
                           'Response text: {}').format(self.channel,
                                                       response.content)
                self._logger.error(message)
                result = False
            else:
                self._logger.debug('Message sent successfully.')
                result = True
        except Exception:
            self._logger.exception()
            return False
        return result


notifierlib/channels/jabber.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
# File: jabber.py

import sleekxmpp
import logging
from notifierlib.notifierlib import Channel

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>, Argiris Gounaris <agounaris@gmail.com>'''
__docformat__ = 'plaintext'
__date__ = '''19-09-2017'''


class XmppClient(sleekxmpp.ClientXMPP):
    """A basic SleekXMPP bot, logs in, sends message, logs out."""

    def __init__(self,
                 user_id,
                 password,
                 recipient,
                 message,
                 server,
                 port,
                 tls=False,
                 ssl=True,
                 reattempt=False):
        super(XmppClient, self).__init__(user_id, password)
        self._logger = logging.getLogger(self.__class__.__name__)
        self.recipient = recipient
        self.message = message
        self.server = server
        self.port = port
        self.tls = tls
        self.ssl = ssl
        self.reattempt = reattempt
        self.add_event_handler('session_start', self.start)
        self.register_plugin('xep_0030')  # Service Discovery
        self.register_plugin('xep_0199')  # XMPP Ping
        # Connect to the XMPP server and start processing XMPP stanzas.
        if not self.connect((self.server, self.port),
                            use_tls=self.tls,
                            use_ssl=self.ssl,
                            reattempt=self.reattempt):
            message = ('Could not connect to '
                       '{server}:{port}').format(server=self.server,
                                                 port=self.port)
            self._logger.error(message)
            raise SyntaxError(message)
        self.process(block=True)

    def start(self, event):
        _ = event  # noqa
        self.send_message(mto=self.recipient,
                          mbody=self.message,
                          mtype='chat')
        self.disconnect(wait=True)


class XmppGroupClient(sleekxmpp.ClientXMPP):
    """A basic SleekXMPP bot, logs in, sends message, logs out."""

    def __init__(self,
                 user_id,
                 password,
                 room,
                 nickname,
                 message,
                 server,
                 port,
                 room_password=None,
                 tls=False,
                 ssl=True,
                 reattempt=False):
        super(XmppGroupClient, self).__init__(user_id, password)
        self._logger = logging.getLogger(self.__class__.__name__)
        self.room = room
        self.room_password = room_password
        self.nickname = nickname
        self.message = message
        self.server = server
        self.port = port
        self.tls = tls
        self.ssl = ssl
        self.reattempt = reattempt
        self.add_event_handler('session_start', self.start)
        self.register_plugin('xep_0030')  # Service Discovery
        self.register_plugin('xep_0045')  # Multi-User Chat
        self.register_plugin('xep_0199')  # XMPP Ping
        # Connect to the XMPP server and start processing XMPP stanzas.
        if not self.connect((self.server, self.port),
                            use_tls=self.tls,
                            use_ssl=self.ssl,
                            reattempt=self.reattempt):
            message = ('Could not connect to '
                       '{server}:{port}').format(server=self.server,
                                                 port=self.port)
            self._logger.error(message)
            raise SyntaxError(message)
        self.process(block=True)

    def start(self, event):
        _ = event  # noqa
        self.plugin['xep_0045'].joinMUC(self.room,
                                        self.nickname,
                                        # If a room password is needed, use:
                                        password=self.room_password,
                                        wait=True)
        self.send_message(mto=self.room,
                          mbody=self.message,
                          mtype='groupchat')
        self.disconnect(wait=True)


class Jabber(Channel):

    def __init__(self,
                 name,
                 user_id,
                 password,
                 recipient_id,
                 server,
                 port,
                 tls=False,
                 ssl=True,
                 reattempt=False):
        super(Jabber, self).__init__(name)
        self._logger = logging.getLogger(self.__class__.__name__)
        self.user = user_id
        self.password = password
        self.server = server
        self.recipient = recipient_id
        self.port = port
        self.tls = tls
        self.ssl = ssl
        self.reattempt = reattempt

    def notify(self, **kwargs):
        message = kwargs.get('message')
        try:
            _ = XmppClient(self.user,  # noqa
                           self.password,
                           self.recipient,
                           message,
                           self.server,
                           self.port,
                           self.tls,
                           self.ssl,
                           self.reattempt)
        except Exception:
            self._logger.exception()
            return False
        return True


class JabberGroup(Channel):
    def __init__(self,
                 name,
                 user_id,
                 password,
                 room,
                 nickname,
                 server,
                 port,
                 room_password=None,
                 tls=False,
                 ssl=True,
                 reattempt=False):
        super(JabberGroup, self).__init__(name)
        self._logger = logging.getLogger(self.__class__.__name__)
        self.user = user_id
        self.password = password
        self.nickname = nickname
        self.room = room
        self.room_password = room_password
        self.server = server
        self.port = port
        self.tls = tls
        self.ssl = ssl
        self.reattempt = reattempt

    def notify(self, **kwargs):
        message = kwargs.get('message')
        try:
            _ = XmppGroupClient(self.user,  # noqa
                                self.password,
                                self.room,
                                self.nickname,
                                message,
                                self.server,
                                self.port,
                                self.room_password,
                                self.tls,
                                self.ssl,
                                self.reattempt)
        except Exception:
            self._logger.exception()
            return False
        return True


notifierlib/channels/stdout.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
# File: stdout.py


from notifierlib.notifierlib import Channel


__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = 'plaintext'
__date__ = '''18-09-2017'''


class Stdout(Channel):
    """A simple library to print to stdout"""

    def __init__(self, name):
        super(Stdout, self).__init__(name)

    @staticmethod
    def _format_arguments(arguments):
        if not arguments.keys():
            return ''
        output = []
        longer_string_length = len(max(arguments.keys(), key=len))
        keys = sorted(arguments.keys())
        for entry in keys:
            width = longer_string_length + 1 - len(entry)
            text = ' '.join([word.capitalize() for word in entry.split()])
            value = arguments[entry]
            output.append('{legend:{align}{width}} :{value}'.format(legend=text,
                                                                    align='<',
                                                                    width=width,
                                                                    value=value)
                          )
        return '\n'.join(output)

    def notify(self, **kwargs):
        try:
            print(self._format_arguments(kwargs))
        except Exception:
            self._logger.exception()
            return False
        return True


notifierlib/channels/telegram_.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
# File: telegram_.py

import telegram
import logging
from notifierlib.notifierlib import Channel
from jinja2 import Environment as Env


__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = 'plaintext'
__date__ = '''19-09-2017'''


class Telegram(Channel):
    def __init__(self, name, token, chat_id, template=None, formatting=None):
        super(Telegram, self).__init__(name)
        self.chat_id = chat_id
        self.template = template
        self.formatting = self._get_formatting(formatting)
        self._bot = telegram.Bot(token)

    @staticmethod
    def _get_formatting(formatting):
        if formatting:
            if formatting.upper() not in ['MARKDOWN', 'HTML']:
                raise ValueError('Unsupported formatting {}'.format(formatting))
            else:
                formatting = formatting.upper()
        return formatting

    def notify(self, **kwargs):
        try:
            body = Env().from_string(self.template).render(**kwargs) if \
                self.template else kwargs.get('message')
            arguments = {'chat_id': self.chat_id,
                         'text': body}
            if self.formatting:
                parse_mode = getattr(telegram.ParseMode, self.formatting)
                arguments['parse_mode'] = parse_mode
            self._bot.send_message(**arguments)
        except Exception:
            self._logger.exception()
            return False
        return True


notifierlib/channels/mattermost.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
# File: mattermost.py


import logging
import requests
import json
from notifierlib.notifierlib import Channel
from jinja2 import Environment

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = 'plaintext'
__date__ = '''19-09-2017'''


class Mattermost(Channel):

    def __init__(self, name, webhook_url, template=None):
        super(Mattermost, self).__init__(name)
        self.url = webhook_url
        self.template = template

    def notify(self, **kwargs):
        try:
            if self.template:
                body = Environment().from_string(self.template).render(**kwargs)
            else:
                body = kwargs.get('message')
            headers = {'Content-Type': 'application/json'}
            payload = {'text': body}
            response = requests.post(self.url,
                                     data=json.dumps(payload),
                                     headers=headers,
                                     timeout=20)
            if not response.ok:
                message = ('Error sending message to {url}.\n'
                           'Message: {message}\n'
                           'Response: '
                           '{response}\n').format(url=self.url,
                                                  message=body,
                                                  reason=response.content)
                self._logger.error(message)
                return False
            self._logger.debug(('Message sent successfully. Response text: '
                                '{response}').format(response=response.text))
        except Exception:
            self._logger.exception()
            return False
        return True


notifierlib/channels/email.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
# File: email.py
"""Email module file"""

from emaillib import EasySender
from notifierlib.notifierlib import Channel
from jinja2 import Environment

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = 'plaintext'
__date__ = '''18-09-2017'''


class Email(Channel):
    def __init__(self,
                 name,
                 sender,
                 recipient,
                 smtp_address,
                 username=None,
                 password=None,
                 tls=False,
                 ssl=True,
                 port=587,
                 template=None,
                 content='text'):
        super(Email, self).__init__(name)
        self.recipient = recipient
        self.sender = sender
        self.email = EasySender(smtp_address=smtp_address,
                                username=username,
                                password=password,
                                ssl=ssl,
                                tls=tls,
                                port=port)
        self.template = template
        self.content = content

    def notify(self, **kwargs):
        try:
            if self.template:
                body = Environment().from_string(self.template).render(**kwargs)
            else:
                body = kwargs.get('message')
            result = self.email.send(sender=self.sender,
                                     recipients=self.recipient,
                                     subject=kwargs.get('subject'),
                                     body=body,
                                     content=self.content)
            if not result:
                self._logger.error('Failed sending email')
        except Exception:
            self._logger.exception()
            return False
        return True if result else False
